kubectl: Verify dry run support

apelisse opened the pull request:
We want to verify if the resource supports dry-run before we send dry-run queries to it, because old servers are just ignoring the query-param and will persist the objects. Not good. 

Even though the heuristic isn't perfect yet, I think it's obviously better than what we have today.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

Commit:
Create a method to check dryRun is supported

We don't want to run dryRun requests against servers that don't support
dry-run, since they might ignore the flag and just persist the unwanted
changes.

This creates a new method that checks in the OpenAPI if the dryRun
parameter can be used.

Commit:
dry-run: Verify if the object supports dry-run

For each object, first we verify if they have a dryRun parameter in the
openapi for the patch verb. If we can't find the object, we assume that
CRD will behave like "namespace". So we check if namespace supports
dryRun. If it does, then we verify that the resource is a CRD.

apelisse commented:
/kind bug

seans3 commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -684,6 +693,22 @@ type patcher struct {
 
 func (p *patcher) patchSimple(obj runtime.Object, modified []byte, source, namespace, name string, errOut io.Writer) ([]byte, runtime.Object, error) {
+	if p.serverDryRun {
+		doc, err := p.discoveryClient.OpenAPISchema()
```
patchSimple is already a large function. Should we break this check into it's own patcher method? Additionally, can we add significant comments to describe why this is necessary (what bug this fixes).

liggitt commented on pkg/kubectl/cmd/util/openapi/dryrun.go:
```diff
@@ -0,0 +1,65 @@
+
+func checkExtension(extensions []*openapi_v2.NamedAny, gvk schema.GroupVersionKind) bool {
+	for _, extension := range extensions {
+		if extension.GetValue().GetYaml() == "" ||
```
I'm not familiar with this method of getting data out of the openapi doc... it's yaml-encoded?

cc @mbohlool for review

apelisse replied:
Since vendor extensions can be any type of yaml data in the openapi spec, this framework is just keeping it as bytes that can be decoded into whatever needed.

mbohlool replied:
Antonie used gnostic openapi structs more than me, if this works, I am fine with it. If I want to be on the safe side, I will only check the name and if the value is empty, log error or fail or something because that is not usual.

mbohlool commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -683,7 +692,38 @@ type patcher struct {
+// supports dryRun. Sending dryRun requests to apiserver that don't
+// support it will result in objects being unwillingly persisted.
+//
+// If the GVK can not be found (it can be a CRD or a resource coming
```
I won't do this assumption specially for aggregated server. I would say it safer to say it does not support dry-run than assuming it does.

apelisse replied:
Agreed. I guess my argument is that this piece of code is safer than what we have today (nothing is checked at all).

mbohlool commented on pkg/kubectl/cmd/util/openapi/dryrun.go:
```diff
@@ -0,0 +1,65 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+func checkExtension(extensions []*openapi_v2.NamedAny, gvk schema.GroupVersionKind) bool {
```
I would use a better name here.e.g., `hasGVKExtension`?

apelisse replied:
Good suggestion, thank you!

Commit:
dry-run: Create class to find if a type is a CRD

Finding out if a Group-version-kind is a CRD is useful, since we want to
detect dry-run ability differently for CRDs.

apelisse commented:
I've been caught by the gofmt change! :-) Fixed all the things.

seans3 commented:
/retest
/lgtm

seans3 commented:
/retest
/lgtm
